service/history/workflow/mutable_state_impl.go

```diff
@@ -1931,13 +1942,23 @@ func (ms *MutableStateImpl) ApplyWorkflowExecutionStartedEvent(
 	if event.SearchAttributes != nil {
 		ms.executionInfo.SearchAttributes = event.SearchAttributes.GetIndexedFields()
 	}
+	// TODO: [cleanup-old-wv]
 	if event.SourceVersionStamp.GetUseVersioning() && event.SourceVersionStamp.GetBuildId() != "" {
 		limit := ms.config.SearchAttributesSizeOfValueLimit(string(ms.namespaceEntry.Name()))
 		if _, err := ms.addBuildIdsWithNoVisibilityTask([]string{worker_versioning.VersionedBuildIdSearchAttribute(event.SourceVersionStamp.BuildId)}, limit); err != nil {
 			return err
 		}
 	}
 
+	if inheritedBuildId := event.InheritedBuildId; inheritedBuildId != "" {
```

Comment:
I think this should be an `else if` to be clear these are exclusive options. or else combine more of the logic with the case above